bumble/pandora/l2cap.py

```diff
@@ -0,0 +1,318 @@
+    ) -> l2cap_pb2.ConnectResponse:
+        self.log.info('Connect')
+        channel: Union[
+            FixedChannelProxy, l2cap.ClassicChannel, l2cap.LeCreditBasedChannel
+        ]
+        connection_handle = int.from_bytes(request.connection.cookie.value, 'big')
+
+        connection = self.device.lookup_connection(connection_handle)
+        if connection is None:
+            raise RuntimeError('Connection not exist')
+
+        if request.type_variant() == 'fixed':
+            # For fixed channel connection, do nothing because it's connectionless
+            assert request.fixed
+            cid = request.fixed.cid
+            l2cap_cookie = any_pb2.Any(value=struct.pack('>HH', connection_handle, cid))
+            self.channels[connection_handle][cid] = FixedChannelProxy(
+                connection_handle=connection_handle,
+                cid=cid,
+                device=self.device,
+            )
+
+            def on_fixed_pdu(connection_handle: int, pdu: bytes) -> None:
+                self.channels[connection_handle][cid].on_data(pdu)
+
+            self.device.l2cap_channel_manager.register_fixed_channel(cid, on_fixed_pdu)
+            return l2cap_pb2.ConnectResponse(
+                channel=l2cap_pb2.Channel(cookie=l2cap_cookie)
+            )
+
+        if request.type_variant() == 'basic':
+            assert request.basic
+            channel = await connection.create_l2cap_channel(
+                spec=l2cap.ClassicChannelSpec(
+                    psm=request.basic.psm, mtu=request.basic.mtu
+                )
+            )
+        elif request.type_variant() == 'le_credit_based':
+            assert request.le_credit_based
+            channel = await connection.create_l2cap_channel(
+                spec=l2cap.LeCreditBasedChannelSpec(
+                    psm=request.le_credit_based.spsm,
+                    max_credits=request.le_credit_based.initial_credit,
+                    mtu=request.le_credit_based.mtu,
+                    mps=request.le_credit_based.mps,
+                )
+            )
+        else:
+            raise NotImplementedError
+
+        self.channels[connection_handle][channel.source_cid] = CocChannelProxy(channel)
+        l2cap_cookie = any_pb2.Any(
+            value=struct.pack('>HH', connection_handle, channel.source_cid)
+        )
+        return l2cap_pb2.ConnectResponse(channel=l2cap_pb2.Channel(cookie=l2cap_cookie))
+
+    @utils.rpc
+    async def OnConnection(
+        self, request: l2cap_pb2.OnConnectionRequest, context: grpc.ServicerContext
+    ) -> AsyncGenerator[l2cap_pb2.OnConnectionResponse, None]:
+        self.log.info('WaitConnection')
+
+        queue: asyncio.Queue[l2cap_pb2.OnConnectionResponse] = asyncio.Queue()
+        connection_handle = int.from_bytes(request.connection.cookie.value, 'big')
+
+        connection = self.device.lookup_connection(connection_handle)
+        if connection is None:
+            raise RuntimeError('Connection not exist')
```

Comment:
Isnt the `connection` field optional ?

Reply:
I thought you don't want to support this feature?
https://github.com/google/bt-test-interfaces/pull/8#discussion_r1282284692

Reply:
I apologize for this because I can only admit that I was wrong. The reason I forgot about at first was: For timing issue we need to be able listen for incoming L2CAP channel connection before ACL's connect. I'm wondering if we need to change the APIs, WDYT ? @DeltaEvo ? Also, we need to remove the `fixed` field from the `ConnectRquest`

Reply:
You may mark it optional or totally remove it.